code/network/lib/Loot/Network/ZMQ.hs

```diff
@@ -30,12 +28,14 @@ Client:
   and propagates them to backend, which sends them to the network.
   Broker also gets messages from the PULL and propagates them to
   a worker that "subscribed" to this kind of update.
-
+* Every time we need to connect to some server, we open a DEALER socket,
+  ask server's ROUTER for his zmq identity, disconnect and destroy this
+  DEALER and use ROUTER <-> ROUTER communication.
```

Comment:
Btw could you explain why so? I was wondering yet when I read the whole code.

Reply:
In distributed ZMQ networks you basically have two options: ROUTER <-> ROUTER and DEALER <-> ROUTER. You should understand that the second option means that the client has a DEALER for each peer it's connected to. This is more or less what we're doing at the "connection handshake" step, right. Both approaches have advantages and disadvantages. The most straightforward advantage of DEALER <-> ROUTER is that we don't care about identifiers as whenever we get the message from peer P we receive it on the socket related to P, so the whole idea of identifiers can be ditched. The issue is that in our networking prototype we also have PUB <-> SUB and we really want updates to have node identity inside (so thus we can query the block with hash H from N when we get update H from the node N). So in order to solve this problem in case we use DEALER <-> ROUTER we can either use identities (but then why would we switch to DEALER <-> ROUTER) or introduce a SUB socket per peer -- which I find somewhat counterintuitive as all the PUB/SUB manuals recommend using one SUB for many PUBs.